feat(server, web): include pictures of shared albums on map

andreasgerstmayr opened the pull request:
This PR adds a new setting "Include shared albums" to the map settings, to includes pictures of albums shared with the current user on the map.

Related: #3176

andreasgerstmayr committed:
feat(server, web): include pictures of shared albums on map

andreasgerstmayr commented on server/src/infra/repositories/asset.repository.ts:
```diff
@@ -511,19 +511,35 @@ export class AssetRepository implements IAssetRepository {
-        exifInfo: {
-          latitude: Not(IsNull()),
-          longitude: Not(IsNull()),
+      where: [
```
Is there a better way to write this OR expression?
`WHERE (ownerID IN ... OR albumId IN ...) AND isVisible = ... AND isArchived = ... AND ...`

andreasgerstmayr committed:
run prettier

andreasgerstmayr committed:
Merge remote-tracking branch 'upstream/main' into include-shared-albums-in-map

andreasgerstmayr committed:
re-create api clients

jrasm91 commented on server/src/domain/asset/asset.service.ts:
```diff
@@ -159,14 +161,19 @@ export class AssetService {
     }
-    return this.assetRepository.getMapMarkers(userIds, options);
+    if (options.withSharedAlbums) {
+      const sharedAlbums = await this.albumRepository.getShared(auth.user.id);
```
This doesn't include the asset relation does it?

andreasgerstmayr replied:
I don't follow, where should I include the asset relation?

NicholasFlamy replied:
Okay so Jason answered me on Discord about this:
https://discord.com/channels/979116623879368755/994044917355663450/1238973376349864108

Jason said: "Just want to make sure this is not loading the album AND all the assets for it"

I said: "If I understand correctly, you are asking if it is loading only the shared albums when figuring out what to show on the map, versus loding the shared albums and all the assets. And the problem with that is loading all the assets in the shared album takes more resources."

Jason responded: "Yup"

andreasgerstmayr replied:
yes, this function does not query all assets, only the albums:
https://github.com/immich-app/immich/blob/f667c9597b9885059de00d9e8779147c4d08ddc2/server/src/repositories/album.repository.ts#L154-L166

jrasm91 commented on server/src/domain/asset/dto/map-marker.dto.ts:
```diff
@@ -31,4 +31,10 @@ export class MapMarkerDto {
   @Transform(toBoolean)
   withPartners?: boolean;
+
+  @ApiProperty()
```
This should not be needed

andreasgerstmayr replied:
:+1:

The class got moved to `server/src/dtos/search.dto.ts` and doesn't have this annotation anymore.

jrasm91 commented on server/src/infra/repositories/asset.repository.ts:
```diff
@@ -511,19 +515,35 @@ export class AssetRepository implements IAssetRepository {
+          albums: {
+            id: In([...albumIds]),
+          },
+          isVisible: true,
+          isArchived,
+          exifInfo: {
+            latitude: Not(IsNull()),
+            longitude: Not(IsNull()),
+          },
+          isFavorite,
+          fileCreatedAt: OptionalBetween(fileCreatedAfter, fileCreatedBefore),
```
This section is repeated twice. Can you move it up to a variable?

andreasgerstmayr replied:
Good idea, done.

jrasm91 commented on web/src/lib/components/map-page/map-settings-modal.svelte:
```diff
@@ -35,6 +35,7 @@
       <SettingSwitch title="Only favorites" bind:checked={settings.onlyFavorites} />
       <SettingSwitch title="Include archived" bind:checked={settings.includeArchived} />
       <SettingSwitch title="Include shared with me" bind:checked={settings.withPartners} />
+      <SettingSwitch title="Include shared albums" bind:checked={settings.withSharedAlbums} />
```
Should we remember this other option so it is more obvious what the difference is?

andreasgerstmayr replied:
I renamed `Include shared with me` to `Include shared from partners`. Another option would be `Include assets from partners`.

What do you think, which option sounds better? Or do you have other suggestions?

NicholasFlamy replied:
I'm not Jason but I would suggest `Include partner assets` or `Include shared partner assets`. Also, this rename is important for clarity.

andreasgerstmayr replied:
I renamed it to `Include shared partner assets` now, thanks for the suggestion @NicholasFlamy

jrasm91 reviewed (approved):
This looks really good. Thanks!

danieldietzler commented:
@andreasgerstmayr You've requested a review from me but you haven't yet addressed @jrasm91 comments and there are also lots of merge conflicts right now...?

andreasgerstmayr commented:
> @andreasgerstmayr You've requested a review from me but you haven't yet addressed @jrasm91 comments and there are also lots of merge conflicts right now...?

hi @danieldietzler, sorry about that, must have been some automated process (it's showing up as 2:35 in the night), maybe because you're listed as a codeowner. No idea why it's happening now and not when I created the PR though.

Sorry for the late response, I was on holiday for a couple of weeks and then didn't get time so far.
Thanks for the review, I'll rebase and implement the changes soon.

danieldietzler commented:
Ah ok got it. No rush then :)

andreasgerstmayr committed:
implement suggestions from code review

andreasgerstmayr committed:
Merge remote-tracking branch 'upstream/main' into include-shared-albums-in-map

andreasgerstmayr committed:
shared from partner -> shared from partners

andreasgerstmayr commented:
I fixed the merge conflicts and implemented (most of) @jrasm91 suggestions now (except the one I didn't understand, see above).
Sorry for the delay on my side!

andreasgerstmayr committed:
rename to 'include shared partner assets'

zackpollard reviewed (approved):
LGTM, will let @jrasm91 take a look over it before we merge though

zackpollard committed:
Merge branch 'refs/heads/main' into fork/include-shared-albums-in-map

zackpollard committed:
chore: fix tsc error in server and prettier in web

zackpollard commented:
I fixed a tsc error caused by changes in main since the branch was created as well as a prettier error in web. The PR should now be good to merge

jrasm91 committed:
fix: include assets shared via owner albums

jrasm91 reviewed (approved):
I updated this to include assets that are in shared albums where the user is the owner.

andreasgerstmayr commented:
> I updated this to include assets that are in shared albums where the user is the owner.

Good catch! I didn't think of that.

Thank you for merging! :)
